Add Python 3.13 support and fix enum behavior difference

willingc opened the pull request:
Closes #237.  This PR adds Python 3.13 support.

Co-authored with @psobolewskiPhD using #240 code changes for CI.

## Key Changes

  1. Enum Compatibility Layer (`napari_animation/_enum_compat.py`):
    - Python 3.13 introduced stricter enum handling requiring `enum.member()` wrapper for callable values
    - Created compatibility layer that detects Python version and conditionally wraps enum values
    - Updated Easing and Interpolation enums to use this wrapper
  2. CI/CD Updates:
    - Added Python 3.13 to test matrix
    - Updated GitHub Actions from v4 to v5
  3. Development Tool Updates:
    - Black updated to v25.1.0 with py313 target
    - Added skip-string-normalization to preserve quote styles
    - Fixed ruff linting and Black formatting issues
  4. Test Infrastructure:
    - Added comprehensive enum compatibility tests
    - Updated mocking for Python 3.10 compatibility
  5. Drop support for 3.9

  The changes ensure napari-animation works seamlessly across Python 3.10-3.13 while maintaining backward compatibility.

psobolewskiPhD commented on napari_animation/_enum_compat.py:
```diff
@@ -0,0 +1,37 @@
+This compatibility layer detects Python version and conditionally wraps enum values.
+- Python 3.10:
+- Python 3.11 added `member`
+- Python 3.13+: No wrapper required.
```
Yes, that is my understanding as well. py313 requires using enum.member, but py310 doesn't have that hence the wrapper being needed.
Once we drop 3.10, then we won't need the compatibility layer and can just use `member(partial(...))` from `enum`

jni replied:
Should we just drop 3.10 then? [SPEC0 says yes](https://scientific-python.org/specs/spec-0000/#support-window).

jni replied:
A fair point! I'm happy to leave it. You're making me want to follow SPEC0 more strictly on napari main, though 😜. 

I think though that we should be even clearer about the bits of code that can be removed. (See this discussion for example.) I think the whole `_enum_compat` file should be able to go once we drop 3.10? Which will be in a few months with napari 0.7.0.

psobolewskiPhD commented on napari_animation/_tests/test_enum_compat.py:
```diff
@@ -0,0 +1,190 @@
+    assert result == 10
+
+
+@pytest.mark.skip(reason="Troubleshooting")
+def test_wrap_enum_member_python_310_simulation():
```
I don't quite get this test. We run on py310 and py313, so the other tests should cover both behaviors?

psobolewskiPhD commented on napari_animation/_tests/test_enum_compat.py:
```diff
@@ -0,0 +1,190 @@
+        assert wrapped(5) == 10
+
+
+@pytest.mark.skip(reason="Not implemented yet")
+def test_wrap_enum_member_python_313_simulation():
```
Same as above, I don't get these simulation tests.

psobolewskiPhD commented on tox.ini:
```diff
@@ -1,25 +1,25 @@
 platform = 
-    macos: darwin
-    macosintel: darwin
+    macos_arm64: darwin-arm64
+    macos_x86_64: darwin-x86_64
```
This should be revered, the platform is just `darwin` as far as I know

psobolewskiPhD commented on tox.ini:
```diff
@@ -1,25 +1,25 @@
-    macos-latest: macos
-    macos-13: macosintel
-    windows-latest: windows
+    macos_arm64: darwin-arm64
+    macos_x86_64: darwin-x86_64
+    linux: linux
+    windows: win32
```
I think this needs to be reverted, because this is supposed to map the GitHub runner (`platform`) to the tox env.
The runners are defined in test_and_deploy:
`platform: [ ubuntu-latest, windows-latest, macos-latest ]`
So looks like we dropped testing on macosintel and could just remove that.


Or it needs to be changed to have the GitHub names on the left hand side...

psobolewskiPhD commented on tox.ini:
```diff
@@ -36,20 +36,26 @@ deps =
+[testenv:py{310}-{linux,macos_x86_64,macos_arm64,windows}-pyside6]
+deps =
+    {[testenv]deps}
+    napari[PySide6]
```
this doesn't exist (yet)

psobolewskiPhD commented on tox.ini:
```diff
@@ -36,20 +36,26 @@ deps =
+[testenv:py{310}-{linux,macos_x86_64,windows}-pyside2]
 deps =
     {[testenv]deps}
     napari[pyside2]
```
pyside2 requires `pytest-qt < 4.5.0`

psobolewskiPhD reviewed (dismissed):
I think the enum compatibility is a very clever fix till we drop 3.10! Thanks!
but there are some CI issues with the tox.ini changes.
I've tried to break out just fixing CI and running on py313 into their own PR:
https://github.com/napari/napari-animation/pull/240
Hopefully that makes this PR easier!

jni commented on napari_animation/_enum_compat.py:
```diff
@@ -0,0 +1,37 @@
+def wrap_enum_member(value):
+    """Conditionally wrap a value with enum.member if needed."""
+    if _NEEDS_ENUM_MEMBER:
+        if not _HAS_ENUM_MEMBER:
+            raise RuntimeError(
+                f"Python {sys.version_info[:2]} requires enum.member but it's not available. "
+                "Please upgrade your Python installation."
+            )
```
This code should be unreachable, so I am in favour of removing it (and the corresponding test)

jni commented on napari_animation/_enum_compat.py:
```diff
@@ -0,0 +1,37 @@
+"""Compatibility layer for enum behavior across Python versions.
+
+Python 3.13 introduced stricter enum handling requiring `enum.member()` wrapper for callable values.
```
Can we link the relevant PEPs or Python docs here?

willingc commented:
@jni @psobolewskiPhD @brisvag Maybe we should simplify everything and drop 3.10 early. We can then drop all of the enum guarding which is a PITA to maintain. I would be inclined to also move this to a modern packaging deployment too. Thoughts?

jni commented:
I actually think we don't need to worry *too much* about simul-dropping — if you are on 3.10 you'll simply get a slightly older napari-animation, but it'll work. Right?

willingc commented:
@jni I added back 3.10 for pyside for linux and window (no mac since I couldn't get it to work). I also updated the ci to output in the run name to have the qt_version displayed.

Let me go back (now that CI is green) and check the nested guard that you mention. Honestly, I spent most of my time fighting with tox and CI, as they were not as explicit as I would have expected. Thankfully, @psobolewskiPhD had the fail-fast toggled to false in his PR which ultimately unblocked me.

willingc commented:
@jni Removed the nested check and improved the naming of the guard variables. This should be ready to land. Thanks for the feedback.

psobolewskiPhD commented:
Just a quick note: we really shouldn't drop py310 in this repo, because it will make it incompatible with pyside2. For the time being, we need to keep that for the bundle, until we move everything over to pyside6.
This is a key plugin that needs to work in the bundle.

jni commented:
@psobolewskiPhD the old version of napari-animation would still be available for the bundle *if* the bundle was on 3.10! But it's actually on 3.11!

https://github.com/napari/packaging/pull/189
